tensorflow/compiler/mlir/hlo/lib/Transforms/buffer_reuse.cc

```diff
@@ -0,0 +1,297 @@
+          if (userange.getFirstUseIndex(itemB) <
+              userange.getFirstUseIndex(*insertionPoint))
+            break;
+          ++insertionPoint;
+        }
+        potReuseVector.insert(insertionPoint, itemB);
+      }
+
+      potentialReuseMap.insert(
+          std::pair<Value, SmallVector<Value, 4>>(itemA, potReuseVector));
+    }
+
+    // Replace all uses of the value that is replaced and
+    // delete the DefiningOp.
+    for (auto &reuse : computeActualReuse(potentialReuseMap)) {
+      for (Value reuseValue : reuse.second) {
+        reuseValue.replaceAllUsesWith(reuse.first);
+        reuseValue.getDefiningOp()->erase();
+      }
+    }
+  }
+
+private:
+  /// Check if all uses of item are dominated by the given block.
+  bool dominatesAllUses(Block *block, Value item) {
+    for (OpOperand &operand : item.getUses()) {
+      if (!dominators.dominates(block, operand.getOwner()->getBlock()))
+        return false;
+    }
+    return true;
+  }
+
+  /// Checks if there is a transitive interference between potReuseValue and the
+  /// value that may replace it, we call this value V. potReuses is the vector
+  /// of all values that can potentially be replaced by V. If potReuseValue
+  /// already replaces any other value that is not part of the potReuses vector
+  /// it cannot be replaced by V anymore.
+  bool transitiveInterference(
+      Value potReuseValue, SmallVector<Value, 4> &potReuses,
+      llvm::MapVector<Value, DenseSet<Value>> &actualReuseMap) {
+    return actualReuseMap.find(potReuseValue) != actualReuseMap.end() &&
+           llvm::any_of(actualReuseMap[potReuseValue], [&](Value vReuse) {
+             return !std::count(potReuses.begin(), potReuses.end(), vReuse);
+           });
+  }
+
+  /// Checks if the types of the given values are compatible for a
+  /// replacement.
+  bool checkReuseCompatibility(Value a, Value b) {
+    auto shapedA = a.getType().cast<ShapedType>();
+    auto shapedB = b.getType().cast<ShapedType>();
+
+    // If both types are shaped we can check for equality.
+    if (shapedA.hasStaticShape() && shapedB.hasStaticShape())
+      return a.getType() == b.getType();
+    // If only one of the types is shaped we cannot detect compatibility since
+    // we do not know how the allocation operation behaves on its operands.
+    if (shapedA.hasStaticShape() != shapedB.hasStaticShape())
+      return false;
+
+    // We need the actual alloc operation of both types. For aliases we need
+    // to check for the defining OP of the alias' origin.
+    Operation *defOpA = a.getDefiningOp();
+    Operation *defOpB = b.getDefiningOp();
+
+    // If the alloc method or the number of operands is not the same the types
+    // might not be compatible.
+    if (defOpA->getName() != defOpB->getName() ||
+        defOpA->getNumOperands() != defOpB->getNumOperands())
+      return false;
+
+    // If all operands are equal the types are compatible.
+    for (auto const &pair :
```

Comment:
This is not true. Consider `memref<?x5xi32>` vs. `memref<5x?xi32`. Also, you need to consider the basetype here, too.

As an extension, this could also work on the size of the element type. Not in this PR, please leave a TODO for this.

Reply:
I was a bit imprecise. Please fix the issue with partially static types and also consider the basetype.

An extension to support reuse for elementtypes of the same size can be a TODO.

Reply:
Use `std::equal`?